External/FEXCore/Source/Interface/Core/Dispatcher/Dispatcher.cpp

```diff
@@ -427,226 +427,262 @@ bool Dispatcher::HandleGuestSignal(FEXCore::Core::InternalThreadState *Thread, i
+    if (IsAVXEnabled) {
+      for (size_t i = 0; i < Core::CPUState::NUM_XMMS; i++) {
+        memcpy(&fpstate->_xmm[i], &Frame->State.xmm.avx.data[i][0], sizeof(__uint128_t));
+      }
+      for (size_t i = 0; i < Core::CPUState::NUM_XMMS; i++) {
+        memcpy(&xstate->ymmh.ymmh_space[i], &Frame->State.xmm.avx.data[i][2], sizeof(__uint128_t));
       }
+    } else {
+      memcpy(fpstate->_xmm, Frame->State.xmm.sse.data, sizeof(Frame->State.xmm.sse.data));
+    }
+
+    // FCW store default
+    fpstate->fcw = Frame->State.FCW;
+    fpstate->ftw = Frame->State.FTW;
+
+    // Reconstruct FSW
+    fpstate->fsw =
+      (Frame->State.flags[FEXCore::X86State::X87FLAG_TOP_LOC] << 11) |
+      (Frame->State.flags[FEXCore::X86State::X87FLAG_C0_LOC] << 8) |
+      (Frame->State.flags[FEXCore::X86State::X87FLAG_C1_LOC] << 9) |
+      (Frame->State.flags[FEXCore::X86State::X87FLAG_C2_LOC] << 10) |
+      (Frame->State.flags[FEXCore::X86State::X87FLAG_C3_LOC] << 14);
+
+    // Copy over signal stack information
+    guest_uctx->uc_stack.ss_flags = GuestStack->ss_flags;
+    guest_uctx->uc_stack.ss_sp = GuestStack->ss_sp;
+    guest_uctx->uc_stack.ss_size = GuestStack->ss_size;
+
+    // Apparently RAX is always set to zero in case of badly misbehaving C applications and variadics.
+    Frame->State.gregs[X86State::REG_RAX] = 0;
+    Frame->State.gregs[X86State::REG_RDI] = Signal;
+    Frame->State.gregs[X86State::REG_RSI] = SigInfoLocation;
+    Frame->State.gregs[X86State::REG_RDX] = UContextLocation;
 
-      // FCW store default
-      fpstate->fcw = Frame->State.FCW;
-      fpstate->ftw = Frame->State.FTW;
-
-      // Reconstruct FSW
-      fpstate->fsw =
-        (Frame->State.flags[FEXCore::X86State::X87FLAG_TOP_LOC] << 11) |
-        (Frame->State.flags[FEXCore::X86State::X87FLAG_C0_LOC] << 8) |
-        (Frame->State.flags[FEXCore::X86State::X87FLAG_C1_LOC] << 9) |
-        (Frame->State.flags[FEXCore::X86State::X87FLAG_C2_LOC] << 10) |
-        (Frame->State.flags[FEXCore::X86State::X87FLAG_C3_LOC] << 14);
-
-      // Copy over signal stack information
-      guest_uctx->uc_stack.ss_flags = GuestStack->ss_flags;
-      guest_uctx->uc_stack.ss_sp = GuestStack->ss_sp;
-      guest_uctx->uc_stack.ss_size = GuestStack->ss_size;
-
-      Frame->State.gregs[X86State::REG_RSI] = SigInfoLocation;
-      Frame->State.gregs[X86State::REG_RDX] = UContextLocation;
+    // Set up the new SP for stack handling
+    // The host is required to provide us a restorer.
+    // If the guest didn't provide a restorer then the application should fail with a SIGSEGV.
+    // TODO: Emulate SIGSEGV when the guest doesn't provide a restorer.
+    // TODO: Support the guest's restorer. Required for libunwind.
+    NewGuestSP -= 8;
+    *(uint64_t*)NewGuestSP = SignalReturn;
+    // TODO: Once we support the guest's restorer.
+    // *(uint64_t*)NewGuestSP = (uint64_t)GuestAction->restorer;
```

Comment:
Since there are a couple of disabled lines like this, I'd recommend declaring a `const bool incomplete_guest_restorer_support` variable at the top, so we make sure this code at least compiles and remains somewhat consistent.

Reply:
Used const.